### ISSUE TITLE
Drag behavior powered by pragmatic is not working when `react-dnd` is also present

### ISSUE THREAD

**Issue by jordanwilliams** [atlassian/pragmatic-drag-and-drop#106]
👋🏾 Hello! I work at Slack and I'm exploring adopting pragmatic as our standard library for drag and drop interactions. We currently use both `react-dnd` and `react-beautiful-dnd`. I did a quick prototype internally with pragmatic and it went great. However, I've run into issues getting basic drag behavior to work in some feature code. I believe I've narrowed it down to an interaction with `react-dnd` library. I put some [example](https://codesandbox.io/p/sandbox/drag-drop-debugging-857fph) code on CodeSanbox to demonstrate the issue. It seems that whenever elements are rendered that leverage the `react-dnd` hooks (even if they're rendered in a completely unrelated part of the tree), it prevents the drag behavior from working as expected for pragmatic powered elements. It seems as though `react-dnd` is calling `event.preventDefault` when they receive the `dragstart` event which then cancels the drag behavior.

Ideally we could just get rid of `react-dnd` altogether, but it's just not possible yet since it powers some of our core experiences that we can't easily refactor. I'm curious to see if you all ran into similar issues at Atlassian when adopting pragmatic internally. I believe Alex mentioned in the intro video on Youtube that Atlassian also used `react-dnd` and `react-beautiful-dnd` before switching to pragmatic. I'm also curious to see if anyone else has run into this. I know this is not necessarily an issue with pragmatic itself, though this interaction is something that could block us from adopting it internally. Some initial ideas I had were to try and stop propagation via an `onDragStart` handler, but this seems like it could negatively impact things like `monitorForElements`. Thank you in advance for any help 🙏🏾 !

**Comment by alexreardon** [comment]
How exiting! I'll take a look at this soon

**Comment by alexreardon** [comment]
Pragmatic drag and drop works on a system of registration. It will only control `draggable` elements that are explicitly registered with `draggable({element})`. 

`react-dnd` _currently_ expects to own the drag interaction for all draggable elements (`<div draggable="true">`) in the the `rootElement` for a `<DndProvider>` (which by default is the `window` 😭). 

To get around the problem you are seeing, you will need to set the `rootElement` on the `<DndProvider>` and let `react-dnd` control all the drag and drop in that area. 

[→ Your example - but updated and working](https://codesandbox.io/p/sandbox/pragmatic-dnd-react-dnd-4x9v39?file=%2Fsrc%2FApp.js%3A214%2C1-243%2C2)

Given that `react-dnd` expects to own all draggable elements under it's `rootElement` (not just ones explicitly registered with `drag`) it looks like you cannot co-mingle `react-dnd` and Pragmatic drag and drop within the same element (eg within the same re-orderable list). 

If isolating the parts of your interface in `<DragProvider>` with a relevant `rootElement` is not feasible, let me know. There is likely an option with `patch-package` to get `react-dnd` to ignore `draggable` elements that it does not own.

**Comment by alexreardon** [comment]
Related: https://github.com/react-dnd/react-dnd/issues/3304

**Comment by alexreardon** [comment]
It won't get you unstuck right now, but I thought I should also raise a suggestion to `react-dnd` for it to ignore drag operations on draggable elements that have not been registered with `react-dnd`: https://github.com/react-dnd/react-dnd/issues/3640

**Comment by alexreardon** [comment]
> Ideally we could just get rid of react-dnd altogether, but it's just not possible yet since it powers some of our core experiences that we can't easily refactor.

Our experience is that it is shorter than you might expect to migrate experiences from `react-dnd` to Pragmatic drag and drop (as you have already done the hard work of setting up your drag and drop logic control flow). Could be worth a small spike if you haven't already 😊 

> I'm curious to see if you all ran into similar issues at Atlassian when adopting pragmatic internally.

We didn't run into this particular issue as our experiences using `react-dnd` only used `react-dnd` for a widget and not throughout the page - so we could quickly swap it out completely for Pragmatic drag and drop. We didn't have a situation where we were running both (but we could have solved the issue with the `react-dnd` `rootElement` option above)

**Comment by jordanwilliams** [comment]
@alexreardon thanks so much for the quick and detailed response 🙇🏾 ! I checked out the issue you linked to. It seems like several other folks have been bitten by the `react-dnd` behavior that you described.

Your suggestion to "isolate parts of your experience to be controlled by `react-dnd` and leverage `rootElement` on `DndProvider`" should be feasible! We currently render the `DndProvider` around our entire app. We'll need to move the provider to the various places that leverage `react-dnd` throughout the app instead. One of the reasons (according to code comments from many years ago) we render it around our app today is to avoid having multiple instances of `HTML5Backend` created and set up simultaneously. It's likely we could end up in a situation where we're rendering multiple providers at once (e.g. two separate components rendered at the same time that both support drag and drop), so we may run into something there. I'll look into this when I can and will report back 👀.

An updated version of `react-dnd` that ignores draggable elements that have not been registered with `react-dnd` would be ideal. Though it doesn't seem like there's been any updates in a while. A patched version of the package seems like a good option if we can't get your first suggestion to work. Thank you again 🏆 !

Lastly, I really enjoyed the Syntax episode you did recently and the intro video on youtube. Both were super insightful. We're excited to use Pragmatic!

**Comment by alexreardon** [comment]
I am keen to hear how you go @jordanwilliams and @doeg. It would be awesome if pdnd made it's way into Slack!

I'll close this issue for now. Feel free to create another issue if you wanted to chat more about this, or something else.

Cheers

**Comment by jordanwilliams** [comment]
@alexreardon I'm out on holiday this week but plan to implement one of your suggestions when I return next week. I'm hoping to refactor some of our existing production code to leverage pdnd, so it should be in Slack soon 🤞🏾 !

Closing the issue sounds great 👍🏾. I'll be sure to reach out if anything else comes up, or to let you know when pdnd is in Slack. Thank you again for the help 🙌🏾 !

**Comment by jordanwilliams** [comment]
@alexreardon an update on where things are! I set out to try and isolate the various `react-dnd` powered parts of our app with the `rootElement` pattern you suggested. I initially made the change in our sidebar, and the pragmatic powered drag interaction started working 🙌🏾. But once I started refactoring the other areas of our app, I started seeing regressions. We can have 3 separate areas of our app with their own `DndProvider` at a given time. Whichever provider was rendered first worked great, but drag and drop would just stop working in the other locations. It seemed like all of the providers were using the same drag and drop manager despite having a different `rootElement` value ([DndProvider.tsx#L70-L88](https://github.com/react-dnd/react-dnd/blob/master/packages/react-dnd/src/core/DndProvider.tsx#L70-L88)). I ended up creating and passing my own drag and drop manager to each provider after seeing [DndProvider.tsx#L54-L57](https://github.com/react-dnd/react-dnd/blob/master/packages/react-dnd/src/core/DndProvider.tsx#L54-L57). That got all of the various areas working as expected while still allowing the pragmatic logic to do its own thing. I ended up with the following component 👇🏾.

```ts
export function ReactDndProvider({ children, elementRef }: ReactDndProviderProps) {
	const [options, setOptions] = useState<{ rootElement: Node } | null>(null);

	useEffect(() => {
		if (options) return;

		const element = elementRef.current;
		if (!element) return;

		setOptions({ rootElement: element });
	}, [elementRef, options]);

	const manager = useMemo(() => {
		if (!options) {
			return;
		}

		return createDragDropManager(HTML5Backend, window, options);
	}, [window, options]);

	if (!options || !manager) return null;

	return <DndProvider manager={manager}>{children}</DndProvider>;
}
```

I didn't end up using this in the end though. While most areas were relatively easy to update, there was one situation that was particularly difficult to cleanly isolate. The `react-dnd` hooks cause runtime errors if they're not called within a provider as well. I ended up patching `react-dnd-html5-backend` with the suggestion (https://github.com/react-dnd/react-dnd/issues/3640) you submitted to `react-dnd`. That allowed the element to be dragged. I also added some logic to return early in their `dragover` and `drop` event handlers since they were preventing drops from working [HTML5BackendImpl.tsL643-L651](https://github.com/react-dnd/react-dnd/blob/7c88c37489a53b5ac98699c46a506a8e085f1c03/packages/backend-html5/src/HTML5BackendImpl.ts#L643-L651). Hopefully this will help save someone time down the road 🙏🏾.
